Sources/GRPC/ReadWriteStates.swift

```diff
@@ -57,23 +57,26 @@ enum WriteState {
   /// - Parameter allocator: An allocator to provide a `ByteBuffer` into which the message will be
   ///     written.
   mutating func write(
-    _ message: Message,
+    _ message: Payload,
     allocator: ByteBufferAllocator
   ) -> Result<ByteBuffer, MessageWriteError> {
     switch self {
     case .notWriting:
       return .failure(.cardinalityViolation)
 
     case let .writing(writeArity, contentType, writer):
-      guard let data = try? message.serializedData() else {
+      // Zero is fine: the writer will allocate the correct amount of space.
+      var buffer = allocator.buffer(capacity: 0)
+      
+      do {
+        try message.serialize(into: &buffer)
+      } catch {
         self = .notWriting
         return .failure(.serializationFailed)
       }
-
-      // Zero is fine: the writer will allocate the correct amount of space.
-      var buffer = allocator.buffer(capacity: 0)
+      let data = buffer.readData(length: buffer.readableBytes)!
+      buffer.clear()
       writer.write(data, into: &buffer)
```

Comment:
is there a prettier way of doing this?

Reply:
Yes, but unfortunately we need to rework the writer API slightly. I have a PR up at the moment which adds support for compression which complicates things a little more. For the time being I think making the writer generic over the payload and having it serialize the payload into the buffer is sufficient.